tests/test_export_onnx.py

```diff
@@ -0,0 +1,177 @@
+from detectron2.export.flatten import TracingAdapter
+from detectron2.export.torchscript_patch import patch_builtin_len
+from detectron2.layers import ShapeSpec
+from detectron2.modeling import build_model
+from detectron2.modeling.roi_heads import KRCNNConvDeconvUpsampleHead
+from detectron2.structures import Boxes, Instances
+from detectron2.utils.testing import (
+    get_sample_coco_image,
+    random_boxes,
+)
+
+# TODO: Remove after https://github.com/facebookresearch/detectron2/pull/4291 is merged
+# Import it from detectron2.utils.testing instead
+STABLE_ONNX_OPSET_VERSION = 16
+
+
+# TODO: Remove after https://github.com/facebookresearch/detectron2/pull/4291 is merged
+# Import it from detectron2.utils.testing instead
+def min_torch_version(min_version: str) -> bool:
+    """
+    Returns True when torch's  version is at least `min_version`.
+    """
+    from packaging import version
+    try:
+        import torch
+    except ImportError:
+        return False
+
+    installed_version = version.parse(torch.__version__.split("+")[0])
+    min_version = version.parse(min_version)
+    return installed_version >= min_version
+
+
+# TODO: Remove after https://github.com/facebookresearch/detectron2/pull/4291 is merged
+# Import it from detectron2.utils.testing instead
+def skipIfUnsupportedMinTorchVersion(min_version):
+    """
+    Skips tests for PyTorch versions older than min_version.
+    """
+
+    def skip_dec(func):
+        def wrapper(self):
+            if not min_torch_version(min_version):
+                raise unittest.SkipTest(
+                    f"module 'torch' has __version__ {torch.__version__}"
+                    f", required is: {min_version}"
+                )
+            return func(self)
+        return wrapper
+    return skip_dec
+
+
+@unittest.skipIf(not _check_module_exists("onnx"), "ONNX not installed.")
+class TestONNXTracingExport(unittest.TestCase):
+
+    def testKeypointHead(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.model = KRCNNConvDeconvUpsampleHead(
+                    ShapeSpec(channels=4, height=14, width=14), num_keypoints=17, conv_dims=(4,)
+                )
+
+            def forward(self, x, predbox1, predbox2):
+                inst = [
+                    Instances((100, 100), pred_boxes=Boxes(predbox1)),
+                    Instances((100, 100), pred_boxes=Boxes(predbox2)),
+                ]
+                ret = self.model(x, inst)
+                return tuple(x.pred_keypoints for x in ret)
+
+        model = M()
+        model.eval()
+
+        def gen_input(num1, num2):
+            feat = torch.randn((num1 + num2, 4, 14, 14))
+            box1 = random_boxes(num1)
+            box2 = random_boxes(num2)
+            return feat, box1, box2
+
+        with patch_builtin_len():
+            self._test_model(
+                model,
+                gen_input(1, 2),
```

Comment:
The original unittest https://github.com/facebookresearch/detectron2/commit/db1b0d46edbdbdb0fd975b58a6dc320b780a5938 is the testing the behavior when the model is traced with `gen_input(a, b)`, but then run the trace with `gen_input(c, d)`. The model is expected to work when `(a, b) != (c, d)`.

This unittest does not verify that behavior.

Reply:
Yes, but it really can't test the same thing this PR.
The models returned by either `torch.jit.trace` or `torch.jit.script` are executable, meaning we can run data through them and get a result. `torch.onnx.export` on the other hand, returns an ONNX model which needs an external a runtime for running through data.

Currently, neither PyTorch nor Detectron2 are not runtime engines for ONNX models, so we cannot feed data through the exported model.

But not everything is lost. We can check the ONNX IR and check whether inputs and outputs are dynamic, which indicates that it could handle inputs of different shapes when a proper runtime is used.

https://github.com/facebookresearch/detectron2/pull/4205 will introduce a pipeline which execute the exported ONNX models using ONNX Runtime, allowing such numerical comparison. 

I will add code to check dynamic shapes now and leave a comment to add a numerical test for 2 different batches when ORT is introduced to detectron2

How does that sound?

Reply:
I can accept that this is currently not testable due to lack of dependencies. However I suspect the code is incorrect on the test I described -- I thought onnx tracing will not work for loops but my onnx knowledge is limited.

Knowing the inputs shapes are "dynamic" does not convince me that the output ONNX model is correct. There is a forloop in the original code over the dynamic axes - how does ONNX IR represent that for loop? Could you show something (maybe the IR or the graph?) that demonstrates it will loop for dynamic #iterations, instead of a static #iterations?